### PR TITLE
First part of VirtualFile async conversions

### PR DESCRIPTION
**Pull request by arpad-m**
## Problem

`VirtualFile` does both reading and writing, and it would be nice if both could be converted to async, so that it doesn't have to support an async read path and a blocking write path (especially for the locks this is annoying as none of the lock implementations in std, tokio or parking_lot have support for both async and blocking access).

## Summary of changes

This PR is some initial work on making the `VirtualFile` APIs async. It can be reviewed commit-by-commit.

* Introduce the `MaybeVirtualFile` enum to be generic in a test that compares real files with virtual files.
* Make various APIs of `VirtualFile` async, including `write_all_at`, `read_at`, `read_exact_at`
* Prepare for making the write plus fsync pattern we have in two locations async, but don't do it yet due to data being held that is not `Send`.

Part of #4743

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: pageserver/src/tenant.rs [hunk 1]
```diff
@@ -2421,67 +2424,67 @@ impl Tenant {
...
-        let do_persist = |target_config_path: &Path| -> anyhow::Result<()> {
-            let target_config_parent = target_config_path.parent().with_context(|| {
-                format!(
-                    "Config path does not have a parent: {}",
-                    target_config_path.display()
-                )
-            })?;
+        let do_persist = |target_config_path: PathBuf| -> _ {
+            async move {
+                let target_config_parent = target_config_path.parent().with_context(|| {
```
Comment (problame) [inline pageserver/src/tenant.rs:2437]: Can probably save ourselves one level of indent here?
```suggestion
        let do_persist = |target_config_path: PathBuf| async move {
                let target_config_parent = target_config_path.parent().with_context(|| {
```

File: pageserver/src/tenant/block_io.rs [hunk 2]
```diff
@@ -150,20 +146,19 @@ pub struct FileBlockReader<F> {
     file_id: page_cache::FileId,
 }
 
-impl<F> FileBlockReader<F>
-where
-    F: FileExt,
-{
-    pub fn new(file: F) -> Self {
+impl FileBlockReader<VirtualFile> {
```
Comment (problame) [inline pageserver/src/tenant/block_io.rs:149]: Extracted this stuff into https://github.com/neondatabase/neon/pull/5181

File: pageserver/src/virtual_file.rs [hunk 3]
```diff
@@ -392,10 +346,64 @@ impl Seek for VirtualFile {
...
+    pub fn write_and_fsync(&mut self, buf: &[u8]) -> Result<(), Error> {
+        if self.write(buf)? != buf.len() {
+            return Err(Error::new(
+                std::io::ErrorKind::Other,
+                "Could not write all the bytes in a single call",
+            ));
+        }
+        self.sync_all()?;
+        Ok(())
+    }
```
Comment (problame) [inline pageserver/src/virtual_file.rs:404]: Would prefer to back out the addition of this function. It's not what I had in mind with

>  special-purpose api for write-once files, to be used in save_metadata and persist_tenant_conf

https://github.com/neondatabase/neon/issues/4743#issuecomment-1700663864

Will push a commit that removes this API, then create a PR that adds the API I had in mind.

**Review by problame — COMMENTED** [review]
See my inline comments

**Comment by arpad-m** [comment]
This has been split up into #5181, #5186, #5189, and #5190.

**Pull request closed without merging**